nestor_api/lib/io.py

```diff
@@ -0,0 +1,102 @@
+"""I/O library"""
+
+import errno
+import math
+import os
+import shutil
+import subprocess
+
+from datetime import datetime
+from pathlib import Path
+from random import random
+
+import yaml as yaml_lib
+
+from nestor_api.config.config import Configuration
+
+
+def copy(source, destination):
+    """Copy file or directory with its content from source to destination"""
+    try:
+        shutil.copytree(source, destination)
+    except OSError as err:
+        if err.errno == errno.ENOTDIR:
+            shutil.copy(source, destination)
+        else:
+            raise
+
+
+def create_temporary_directory(directory_prefix=""):
+    """Creates a temporary directory with an optional prefix"""
+    tmp_directory_path = get_temporary_directory_path(directory_prefix)
+
+    # Create directory if it does not exist
+    ensure_dir(tmp_directory_path)
+
+    return tmp_directory_path
+
+
+def ensure_dir(directory_path):
+    """Ensures that a directory exists, else creates it"""
+    Path(directory_path).mkdir(mode=0o777, parents=True, exist_ok=True)
+
+
+def execute(command: str, cwd=None):
+    """Executes a command and returns the stdout from it"""
+    result = subprocess.run(command.split(), stdout=subprocess.PIPE, check=True, cwd=cwd)
+    return result.stdout.decode('utf-8')
+
+
+def exists(file_path):
+    """Checks if a file exists"""
+    return Path(file_path).exists()
+
+
+def get_pristine_path(pristine_path_name):
+    """Returns the pristine path"""
+    return os.path.join(Configuration.pristine_path, pristine_path_name)
+
+
+def get_temporary_copy(directory_path, target_directory_prefix=""):
+    """Creates a copy of a directory in a temporary directory and returns its path"""
+    copy_path = get_temporary_directory_path(target_directory_prefix)
+    copy(directory_path, copy_path)
+
+    return copy_path
+
+
+def get_temporary_directory_path(prefix=""):
+    """Returns a temporary directory path with an optional prefix"""
+    random_num = math.floor(random() * 1e9)
```

Comment:
Why not using native [tempfile.TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) ?

Reply:
Indeed, I did not know, but when I tried out, the issue is that it states the following :

> On completion of the context or **destruction of the temporary directory object** the newly created temporary directory and all its contents are **removed** from the filesystem.

So when I tried, since we only return the path, the object is destroyed and the directory with it 😞

Reply:
Hmmmm okay, no way to make it persistent until we explicitly delete it ?

Reply:
In the doc it doesn't look like, it looks a bit like a smart pointer from cpp.
When no more references to the created object, it is deleted.